Crashed w/ runtime error [Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference]

Hey there 👋🏻

I'm running on AWS EKS cluster v1.25.

The sleepcycles were installed on the cluster and the label were configured on one of the deployments.

Here are some logs from sleepcycles pod:

```
1.70249953358739e+09	INFO	controller-runtime.metrics	Metrics server is starting to listen	{"addr": "127.0.0.1:8080"}
1.7024995335878196e+09	INFO	setup	starting manager
1.7024995335883129e+09	INFO	Starting server	{"kind": "health probe", "addr": "[::]:8081"}
1.7024995335883129e+09	INFO	Starting server	{"path": "/metrics", "kind": "metrics", "addr": "127.0.0.1:8080"}
I1213 20:32:13.588770       1 leaderelection.go:248] attempting to acquire leader lease sleepcycles-system/04cbe4c0.rekuberate.io...
I1213 20:32:30.242666       1 leaderelection.go:258] successfully acquired lease sleepcycles-system/04cbe4c0.rekuberate.io
1.702499550243052e+09	DEBUG	events	Normal	{"object": {"kind":"Lease","namespace":"sleepcycles-system","name":"04cbe4c0.rekuberate.io","uid":"294d077b-9dfd-4795-8120-99bddec3cd2b","apiVersion":"coordination.k8s.io/v1","resourceVersion":"852177480"}, "reason": "LeaderElection", "message": "sleepcycles-controller-manager-bf6c64857-j5k8d_808b4e88-ba76-4555-b558-8d7b48640b11 became leader"}
1.7024995502432828e+09	INFO	Starting EventSource	{"controller": "sleepcycle", "controllerGroup": "core.rekuberate.io", "controllerKind": "SleepCycle", "source": "kind source: *v1alpha1.SleepCycle"}
1.7024995502433164e+09	INFO	Starting Controller	{"controller": "sleepcycle", "controllerGroup": "core.rekuberate.io", "controllerKind": "SleepCycle"}
1.7024995503444824e+09	INFO	Starting workers	{"controller": "sleepcycle", "controllerGroup": "core.rekuberate.io", "controllerKind": "SleepCycle", "worker count": 1}
1.7024995505854793e+09	INFO	📚 Processing Deployments	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown"}
1.7024995508859751e+09	INFO	🕑 Processing CronJobs	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown"}
1.7024995509862914e+09	INFO	📦 Processing StatefulSets	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown"}
1.7024995510869236e+09	INFO	📈 Processing HorizontalPodAutoscalers	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown"}
1.7024995510988872e+09	INFO	🔁 Requeue	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown", "next-op": "Shutdown", "after": "28.901116474s"}
1.7024995511085997e+09	INFO	🔁 Requeue	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown", "next-op": "Shutdown", "after": "28.891404078s"}
1.7024995800032704e+09	INFO	📚 Processing Deployments	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown"}
1.7024995800034337e+09	INFO	🕑 Processing CronJobs	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown"}
1.7024995800034702e+09	INFO	📦 Processing StatefulSets	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown"}
1.702499580003513e+09	INFO	📈 Processing HorizontalPodAutoscalers	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown"}
1.7024995800151165e+09	INFO	🔁 Requeue	{"namespace": "es-digital-journeys", "sleepcycle": "sleepcycle-sample1", "op": "Shutdown", "next-op": "Shutdown", "after": "59.984887885s"}
1.7024995800156264e+09	INFO	Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference	{"controller": "sleepcycle", "controllerGroup": "core.rekuberate.io", "controllerKind": "SleepCycle", "sleepCycle": {"name":"sleepcycle-sample1","namespace":"es-digital-journeys"}, "namespace": "es-digital-journeys", "name": "sleepcycle-sample1", "reconcileID": "f199f8db-368c-43e0-ad11-9c304278acdf"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x13881bb]

goroutine 201 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:118 +0x1f4
panic({0x14c2c80, 0x2328090})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/rekuberate-io/sleepcycles/controllers.(*SleepCycleReconciler).getCurrentScheduledOperation(_, {{{0x13b128d, 0xa}, {0xc001af5780, 0x1b}}, {{0xc001af9b00, 0x12}, {0x0, 0x0}, {0xc001af9b30, ...}, ...}, ...})
	/workspace/controllers/sleepcycle_controller.go:455 +0x17b
github.com/rekuberate-io/sleepcycles/controllers.(*SleepCycleReconciler).Reconcile(0xc000490ab0, {0x18b67c0, 0xc001ba44e0}, {{{0xc0006803a8?, 0x10?}, {0xc000680390?, 0x40d787?}}})
	/workspace/controllers/sleepcycle_controller.go:84 +0x3a5
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x18b6718?, {0x18b67c0?, 0xc001ba44e0?}, {{{0xc0006803a8?, 0x15c5740?}, {0xc000680390?, 0x4041f4?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:121 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001a6b40, {0x18b6718, 0xc00057f4c0}, {0x1512b20?, 0xc0029332c0?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:320 +0x33c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001a6b40, {0x18b6718, 0xc00057f4c0})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:273 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:234 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.2/pkg/internal/controller/controller.go:230 +0x325
```

Here is the output from `**k get sleepcycles.core.rekuberate.io -w**`:

```
NAME                 SHUTDOWN      WAKEUP   ENABLED   NEXT SHUTDOWN          NEXT WAKE UP   LAST RUN               LAST RUN SUCCESS
sleepcycle-sample1   */1 * * * *            true      2023-12-13T20:28:00Z                  2023-12-13T20:27:00Z   true
sleepcycle-sample1   */1 * * * *            true      2023-12-13T20:33:00Z                  2023-12-13T20:32:31Z   true
sleepcycle-sample1   */1 * * * *            true      2023-12-13T20:34:00Z                  2023-12-13T20:33:00Z   true
sleepcycle-sample1   */1 * * * *            true      2023-12-13T20:39:00Z                  2023-12-13T20:38:30Z   true
sleepcycle-sample1   */1 * * * *            true      2023-12-13T20:40:00Z                  2023-12-13T20:39:00Z   true
```

🙏🏻 Thanks you in advance, Let me know in case there is some issue template.

--- a/.gitignore
+++ b/.gitignore
@@ -23,3 +23,5 @@ testbin/*
 *.swp
 *.swo
 *~
+
+kind.yaml

--- a/config/samples/core_v1alpha1_sleepcycle.yaml
+++ b/config/samples/core_v1alpha1_sleepcycle.yaml
@@ -3,6 +3,6 @@ kind: SleepCycle
 metadata:
   name: sleepcycle-sample
 spec:
-  shutdown: "*/3 * * * *"
-  wakeup: "*/4 * * * *"
+  shutdown: "*/1 * * * *"
+  wakeup: "*/2 * * * *"
   enabled: true

--- a/controllers/sleepcycle_controller.go
+++ b/controllers/sleepcycle_controller.go
@@ -20,6 +20,10 @@ import (
 	"context"
 	"fmt"
 	"github.com/go-logr/logr"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"strings"
 	"time"
 
@@ -28,6 +32,7 @@ import (
 	appsv1 "k8s.io/api/apps/v1"
 	autoscalingv1 "k8s.io/api/autoscaling/v1"
 	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
@@ -37,18 +42,42 @@ import (
 )
 
 const (
-	SleepCycleLabel = "rekuberate.io/sleepcycle"
+	SleepCycleLabel                      = "rekuberate.io/sleepcycle"
+	TimeWindowToleranceInSeconds  int    = 30
+	SleepCycleStatusUpdateFailure string = "failed to update SleepCycle Status"
 )
 
 // SleepCycleReconciler reconciles a SleepCycle object
 type SleepCycleReconciler struct {
 	client.Client
-	Scheme *runtime.Scheme
-	logger logr.Logger
+	Scheme   *runtime.Scheme
+	logger   logr.Logger
+	Recorder record.EventRecorder
 }
 
-const (
-	TimeWindowToleranceInSeconds int = 30
+type runtimeObjectReconciler func(
+	ctx context.Context,
+	req ctrl.Request,
+	sleepCycle *corev1alpha1.SleepCycle,
+	deepCopy *corev1alpha1.SleepCycle,
+	op SleepCycleOperation,
+) (ctrl.Result, error)
+
+var (
+	eventFilters = builder.WithPredicates(predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// We only need to check generation changes here, because it is only
+			// updated on spec changes. On the other hand RevisionVersion
+			// changes also on status changes. We want to omit reconciliation
+			// for status updates.
+			return e.ObjectOld.GetGeneration() != e.ObjectNew.GetGeneration()
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// DeleteStateUnknown evaluates to false only if the object
+			// has been confirmed as deleted by the api server.
+			return !e.DeleteStateUnknown
+		},
+	})
 )
 
 //+kubebuilder:rbac:groups=core.rekuberate.io,resources=sleepcycles,verbs=get;list;watch;create;update;patch;delete
@@ -78,15 +107,12 @@ func (r *SleepCycleReconciler) Reconcile(ctx context.Context, req ctrl.Request) 
 		return ctrl.Result{}, err
 	}
 
-	thisRunIsSuccessful := true
-	lastRunWasSuccessful := sleepCycle.Status.LastRunWasSuccessful
-	isScheduledReentry := !r.isNotScheduledReentry(sleepCycle)
-	currentOperation := r.getCurrentScheduledOperation(sleepCycle)
-	sleepCycleFullName := fmt.Sprintf("%v/%v", sleepCycle.Namespace, sleepCycle.Name)
-
 	if !sleepCycle.Spec.Enabled {
 		return ctrl.Result{}, nil
 	}
+
+	currentOperation := r.getCurrentScheduledOperation(sleepCycle)
+	sleepCycleFullName := fmt.Sprintf("%v/%v", sleepCycle.Namespace, sleepCycle.Name)
 
 	deepCopy := *sleepCycle.DeepCopy()
 	if deepCopy.Status.UsedBy == nil {
@@ -96,55 +122,48 @@ func (r *SleepCycleReconciler) Reconcile(ctx context.Context, req ctrl.Request) 
 
 	r.logger = r.logger.WithValues("op", currentOperation.String())
 
-	if isScheduledReentry || !lastRunWasSuccessful {
-		var err error
+	if currentOperation != Watch {
+		reconcilers := []runtimeObjectReconciler{r.ReconcileDeployments, r.ReconcileCronJobs, r.ReconcileStatefulSets, r.ReconcileHorizontalPodAutoscalers}
 
-		_, err = r.ReconcileDeployments(ctx, req, &sleepCycle, &deepCopy, currentOperation)
-		if err != nil {
-			thisRunIsSuccessful = false
+		for _, reconciler := range reconcilers {
+			result, err := reconciler(ctx, req, &sleepCycle, &deepCopy, currentOperation)
+			if err != nil {
+				if currentOperation != Watch {
+					deepCopy.Status.LastRunTime = &metav1.Time{Time: time.Now()}
+					deepCopy.Status.LastRunWasSuccessful = false
+				}
+
+				if err := r.Status().Update(ctx, &deepCopy); err != nil {
+					r.logger.Error(err, fmt.Sprintf("🛑️ %s", SleepCycleStatusUpdateFailure), "sleepcycle", sleepCycleFullName)
+					r.Recorder.Event(&sleepCycle, corev1.EventTypeWarning, "SleepCycleStatus", strings.ToLower(SleepCycleStatusUpdateFailure))
+					return ctrl.Result{}, err
+				}
+
+				return result, err
+			}
 		}
 
-		_, err = r.ReconcileCronJobs(ctx, req, &sleepCycle, currentOperation)
-		if err != nil {
-			thisRunIsSuccessful = false
-		}
-
-		_, err = r.ReconcileStatefulSets(ctx, req, &sleepCycle, &deepCopy, currentOperation)
-		if err != nil {
-			thisRunIsSuccessful = false
-		}
-
-		_, err = r.ReconcileHorizontalPodAutoscalers(ctx, req, &sleepCycle, &deepCopy, currentOperation)
-		if err != nil {
-			thisRunIsSuccessful = false
-		}
+		deepCopy.Status.LastRunTime = &metav1.Time{Time: time.Now()}
+		deepCopy.Status.LastRunWasSuccessful = true
 	}
 
 	nextScheduledShutdown, nextScheduledWakeup := r.getSchedulesTime(sleepCycle, false)
-	deepCopy.Status.NextScheduledShutdownTime = &metav1.Time{Time: *nextScheduledShutdown}
-
-	if currentOperation != Watch {
-		deepCopy.Status.LastRunTime = &metav1.Time{Time: time.Now()}
-		deepCopy.Status.LastRunWasSuccessful = thisRunIsSuccessful
-	}
-
 	if nextScheduledWakeup != nil {
 		deepCopy.Status.NextScheduledWakeupTime = &metav1.Time{Time: *nextScheduledWakeup}
+	} else {
+		deepCopy.Status.NextScheduledWakeupTime = nil
 	}
+	deepCopy.Status.NextScheduledShutdownTime = &metav1.Time{Time: *nextScheduledShutdown}
 
 	if err := r.Status().Update(ctx, &deepCopy); err != nil {
-		r.logger.Error(err, "🛑️ failed to update SleepCycle Status", "sleepcycle", sleepCycleFullName)
+		r.logger.Error(err, fmt.Sprintf("🛑️ %s", SleepCycleStatusUpdateFailure), "sleepcycle", sleepCycleFullName)
+		r.Recorder.Event(&sleepCycle, corev1.EventTypeWarning, "SleepCycleStatus", strings.ToLower(SleepCycleStatusUpdateFailure))
 		return ctrl.Result{}, err
 	}
 
 	nextOperation, requeueAfter := r.getNextScheduledOperation(sleepCycle, &currentOperation)
 
-	if currentOperation == Watch {
-		r.logger.V(10).Info("🔁 Requeue", "next-op", nextOperation.String(), "after", requeueAfter)
-	} else {
-		r.logger.Info("🔁 Requeue", "next-op", nextOperation.String(), "after", requeueAfter)
-	}
-
+	r.logger.Info("🔁 Requeue", "next-op", nextOperation.String(), "after", requeueAfter)
 	return ctrl.Result{RequeueAfter: requeueAfter}, nil
 }
 
@@ -215,7 +234,7 @@ func (r *SleepCycleReconciler) WatchDeploymentsHandler(o client.Object) []ctrl.R
 // SetupWithManager sets up the controller with the Manager.
 func (r *SleepCycleReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&corev1alpha1.SleepCycle{}).
+		For(&corev1alpha1.SleepCycle{}, eventFilters).
 		Complete(r)
 }
 
@@ -250,11 +269,12 @@ func (r *SleepCycleReconciler) ReconcileDeployments(
 			case Watch:
 			case Shutdown:
 				if deployment.Status.Replicas != 0 {
-					r.logger.Info("⬇  Scale Down Deployment", "deployment", deploymentFullName, "targetReplicas", 0)
+					r.logger.Info("🌙 Scale Down Deployment", "deployment", deploymentFullName, "targetReplicas", 0)
 
 					err := r.ScaleDeployment(ctx, deployment, 0)
 					if err != nil {
 						r.logger.Error(err, "🛑️ Scaling Deployment failed", "deployment", deploymentFullName)
+						r.recordEvent(*sleepCycle, true, deploymentFullName, op)
 						return ctrl.Result{}, err
 					}
 				}
@@ -262,11 +282,12 @@ func (r *SleepCycleReconciler) ReconcileDeployments(
 				targetReplicas := int32(deepCopy.Status.UsedBy[deploymentFullName])
 
 				if deployment.Status.Replicas != targetReplicas {
-					r.logger.Info("⬆  Scale Up Deployment", "deployment", deploymentFullName, "targetReplicas", targetReplicas)
+					r.logger.Info("☀️ Scale Up Deployment", "deployment", deploymentFullName, "targetReplicas", targetReplicas)
 
 					err := r.ScaleDeployment(ctx, deployment, targetReplicas)
 					if err != nil {
 						r.logger.Error(err, "🛑️ Scaling Deployment failed", "deployment", deploymentFullName)
+						r.recordEvent(*sleepCycle, true, deploymentFullName, op)
 						return ctrl.Result{}, err
 					}
 				}
@@ -280,6 +301,7 @@ func (r *SleepCycleReconciler) ReconcileDeployments(
 func (r *SleepCycleReconciler) ReconcileCronJobs(ctx context.Context,
 	req ctrl.Request,
 	sleepCycle *corev1alpha1.SleepCycle,
+	deepCopy *corev1alpha1.SleepCycle,
 	op SleepCycleOperation,
 ) (ctrl.Result, error) {
 	cronJobList := batchv1.CronJobList{}
@@ -299,21 +321,23 @@ func (r *SleepCycleReconciler) ReconcileCronJobs(ctx context.Context,
 			case Watch:
 			case Shutdown:
 				if !*cronJob.Spec.Suspend {
-					r.logger.Info("⬇  Suspending CronJob", "cronJob", cronJobFullName)
+					r.logger.Info("🌙 Suspending CronJob", "cronJob", cronJobFullName)
 
 					err := r.SuspendCronJob(ctx, cronJob, true)
 					if err != nil {
 						r.logger.Error(err, "🛑️️ Suspending CronJob failed", "cronJob", cronJobFullName)
+						r.recordEvent(*sleepCycle, true, cronJobFullName, op)
 						return ctrl.Result{}, err
 					}
 				}
 			case WakeUp:
 				if *cronJob.Spec.Suspend {
-					r.logger.Info("⬆  Enabling Cronjob", "cronJob", cronJobFullName)
+					r.logger.Info("☀️ Enabling Cronjob", "cronJob", cronJobFullName)
 
 					err := r.SuspendCronJob(ctx, cronJob, false)
 					if err != nil {
 						r.logger.Error(err, "🛑️️ Suspending CronJob failed", "cronJob", cronJobFullName)
+						r.recordEvent(*sleepCycle, true, cronJobFullName, op)
 						return ctrl.Result{}, err
 					}
 				}
@@ -355,11 +379,12 @@ func (r *SleepCycleReconciler) ReconcileStatefulSets(
 			case Watch:
 			case Shutdown:
 				if statefulSet.Status.Replicas != 0 {
-					r.logger.Info("⬇  Scale Down StatefulSet", "statefulSet", statefulSetFullName, "targetReplicas", 0)
+					r.logger.Info("🌙 Scale Down StatefulSet", "statefulSet", statefulSetFullName, "targetReplicas", 0)
 
 					err := r.ScaleStatefulSet(ctx, statefulSet, 0)
 					if err != nil {
 						r.logger.Error(err, "🛑️ Scaling StatefulSet failed", "statefulSet", statefulSetFullName)
+						r.recordEvent(*sleepCycle, true, statefulSetFullName, op)
 						return ctrl.Result{}, err
 					}
 				}
@@ -367,11 +392,12 @@ func (r *SleepCycleReconciler) ReconcileStatefulSets(
 				targetReplicas := int32(deepCopy.Status.UsedBy[statefulSetFullName])
 
 				if statefulSet.Status.Replicas != targetReplicas {
-					r.logger.Info("⬆  Scale Up StatefulSet", "statefulSet", statefulSetFullName, "targetReplicas", targetReplicas)
+					r.logger.Info("☀️ Scale Up StatefulSet", "statefulSet", statefulSetFullName, "targetReplicas", targetReplicas)
 
 					err := r.ScaleStatefulSet(ctx, statefulSet, targetReplicas)
 					if err != nil {
 						r.logger.Error(err, "🛑️ Scaling StatefulSet failed", "statefulSet", statefulSetFullName)
+						r.recordEvent(*sleepCycle, true, statefulSetFullName, op)
 						return ctrl.Result{}, err
 					}
 				}
@@ -413,11 +439,12 @@ func (r *SleepCycleReconciler) ReconcileHorizontalPodAutoscalers(
 			case Watch:
 			case Shutdown:
 				if hpa.Spec.MaxReplicas != 1 {
-					r.logger.Info("⬇  Scale Down HorizontalPodAutoscaler", "hpa", hpaFullName, "maxReplicas", 1)
+					r.logger.Info("🌙 Scale Down HorizontalPodAutoscaler", "hpa", hpaFullName, "maxReplicas", 1)
 
 					err := r.ScaleHorizontalPodAutoscaler(ctx, hpa, 1)
 					if err != nil {
 						r.logger.Error(err, "🛑️ Scaling HorizontalPodAutoscaler failed", "hpa", hpaFullName)
+						r.recordEvent(*sleepCycle, true, hpaFullName, op)
 						return ctrl.Result{}, err
 					}
 				}
@@ -425,11 +452,12 @@ func (r *SleepCycleReconciler) ReconcileHorizontalPodAutoscalers(
 				targetReplicas := int32(deepCopy.Status.UsedBy[hpaFullName])
 
 				if hpa.Spec.MaxReplicas != targetReplicas {
-					r.logger.Info("⬆  Scale Up HorizontalPodAutoscaler", "hpa", hpaFullName, "maxReplicas", targetReplicas)
+					r.logger.Info("☀️ Scale Up HorizontalPodAutoscaler", "hpa", hpaFullName, "maxReplicas", targetReplicas)
 
 					err := r.ScaleHorizontalPodAutoscaler(ctx, hpa, targetReplicas)
 					if err != nil {
 						r.logger.Error(err, "🛑️ Scaling HorizontalPodAutoscaler failed", "hpa", hpaFullName)
+						r.recordEvent(*sleepCycle, true, hpaFullName, op)
 						return ctrl.Result{}, err
 					}
 				}
@@ -448,7 +476,11 @@ func (r *SleepCycleReconciler) getCurrentScheduledOperation(sleepCycle corev1alp
 	var isWithinScheduleForShutdown, isWithinScheduleForWakeup = false, false
 	isWithinScheduleForShutdown = nextScheduledShutdownTimeWindow.IsScheduleWithinWindow(time.Now())
 
-	if nextScheduledWakeup == nil && isWithinScheduleForShutdown {
+	if nextScheduledWakeup == nil {
+		if !isWithinScheduleForShutdown {
+			return Watch
+		}
+
 		return Shutdown
 	}
 
@@ -563,24 +595,6 @@ func (r *SleepCycleReconciler) getTimeFromCronExpression(cronexp string) *time.T
 	return nil
 }
 
-func (r *SleepCycleReconciler) isNotScheduledReentry(sleepCycle corev1alpha1.SleepCycle) bool {
-	now := metav1.Time{Time: time.Now()}
-
-	if sleepCycle.Status.NextScheduledShutdownTime == nil {
-		return false
-	}
-
-	if now.Time.Before(sleepCycle.Status.NextScheduledShutdownTime.Time) && sleepCycle.Status.NextScheduledWakeupTime == nil {
-		return true
-	}
-
-	if now.Time.Before(sleepCycle.Status.NextScheduledShutdownTime.Time) && (sleepCycle.Status.NextScheduledWakeupTime != nil && now.Before(sleepCycle.Status.NextScheduledWakeupTime)) {
-		return true
-	}
-
-	return false
-}
-
 func (r *SleepCycleReconciler) isTagged(obj *metav1.ObjectMeta, tag string) bool {
 	val, ok := obj.GetLabels()[SleepCycleLabel]
 
@@ -589,4 +603,18 @@ func (r *SleepCycleReconciler) isTagged(obj *metav1.ObjectMeta, tag string) bool
 	}
 
 	return false
+}
+
+func (r *SleepCycleReconciler) recordEvent(sleepCycle corev1alpha1.SleepCycle, isError bool, namespacedName string, operation SleepCycleOperation) {
+	eventType := corev1.EventTypeNormal
+	reason := "SleepCycleOpSuccess"
+	message := fmt.Sprintf("%s on %s succeeded", operation.String(), namespacedName)
+
+	if isError {
+		eventType = corev1.EventTypeWarning
+		reason = "SleepCycleOpFailure"
+		message = fmt.Sprintf("%s on %s failed", operation.String(), namespacedName)
+	}
+
+	r.Recorder.Event(&sleepCycle, eventType, reason, strings.ToLower(message))
 }

--- a/main.go
+++ b/main.go
@@ -90,8 +90,9 @@ func main() {
 	}
 
 	if err = (&controllers.SleepCycleReconciler{
-		Client: mgr.GetClient(),
-		Scheme: mgr.GetScheme(),
+		Client:   mgr.GetClient(),
+		Scheme:   mgr.GetScheme(),
+		Recorder: mgr.GetEventRecorderFor("rekuberate-io/sleepcycles"),
 	}).SetupWithManager(mgr); err != nil {
 		setupLog.Error(err, "unable to create controller", "controller", "SleepCycle")
 		os.Exit(1)

